src/cargo/sources/git/utils.rs

```diff
@@ -944,9 +944,15 @@ pub fn fetch(
 
     let shallow = remote_kind.to_shallow_setting(repo.is_shallow(), gctx);
 
+    // Flag to keep track if the rev is a full commit hash
+    let mut fast_path_rev: bool = false;
+
     let oid_to_fetch = match github_fast_path(repo, remote_url, reference, gctx) {
         Ok(FastPathRev::UpToDate) => return Ok(()),
-        Ok(FastPathRev::NeedsFetch(rev)) => Some(rev),
+        Ok(FastPathRev::NeedsFetch(rev)) => {
+            fast_path_rev = true;
```

Comment:
Instead of adding a flag here, we should probably add one in `} else if let Some(oid_to_fetch) = oid_to_fetch {` after line 995, which the `ref/commit/{0}` really happened and caused the problem